pkg/payforput/payments.go

```diff
@@ -0,0 +1,211 @@
+	Secret string
+}
+
+// New returns a new payment enforcer that can be used for easy BIP70 integration
+// for the keyserver.
+func New(PaymentURL string, Validator ValidatorFunc) *PaymentEnforcer {
+	pe := &PaymentEnforcer{
+		PaymentURL: PaymentURL,
+		Validator:  Validator,
+		Secret:     "",
+	}
+	if pe.Validator == nil {
+		pe.Validator = DefaultValidator
+	}
+	// Generate an ephemeral secret and hold on to it
+	if pe.Secret == "" {
+		pe.Secret = RandString(64)
+	}
+
+	return pe
+}
+
+// DefaultValidator is the default request payment validator
+func DefaultValidator(r *http.Request, secret string) bool {
+	// First attempt to get the code from the querystring
+	token := r.URL.Query().Get("code")
+
+	headerToken := r.Header.Get("Authorization")
+	// If we had an Authorization header, use that instead.
+	if headerToken != "" && len(headerToken) >= 4 && headerToken[0:4] == "POP " {
+		token = headerToken[4:]
+	}
+
+	// Remove the code query param, as it wasn't part of the HMAC hash
+	url := *r.URL
+	url.RawQuery = ""
+	// Validate that the HMAC is valid for this URL.
+	return ValidateHMACToken(url.String(), token, secret)
+}
+
+// PaymentHandler is an http handler that implements a check for payment,
+// along with a redirect to the original location with the key
+func (e *PaymentEnforcer) PaymentHandler(w http.ResponseWriter,
+	r *http.Request) {
+	log := hlog.FromRequest(r)
+	defer r.Body.Close()
+
+	// Check BIP70 headers
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/bitcoin-payment" {
+		http.Error(w, "invalid content type", http.StatusUnsupportedMediaType)
+		return
+	}
+	acceptHeader := r.Header.Get("Accept")
+	if acceptHeader != "application/bitcoin-paymentack" {
+		http.Error(w, "invalid content type", http.StatusUnsupportedMediaType)
+		return
+	}
+
+	// Read the Payment information
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Error().Msgf("unable to read request body: %s", err.Error())
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+	payment := &models.Payment{}
+	err = proto.Unmarshal(body, payment)
+	if err != nil {
+		log.Error().Msgf("unable to unmarshal payment: %s", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+
+	///////
+	// TODO: Check outputs
+	///////
+	log.Info().Str("memo", payment.GetMemo()).Msg("Payment received")
+
+	// Acknowledge the payment, and redirect back to the intended location.
+	memo := "Thank you for being a customer"
+	payAck := &models.PaymentACK{
+		Payment: payment,
+		Memo:    &memo,
+	}
+	payAckBytes, err := proto.Marshal(payAck)
+	if err != nil {
+		log.Error().Msgf("unable to unmarshal payment: %s", err)
+		http.Error(w, "internal server error", http.StatusInternalServerError)
+		return
+	}
+	// Provide a payment token
+	// TODO: Maybe generate something with some metadata
+	token := GenerateHMACToken(string(payment.GetMerchantData()), e.Secret)
```

Comment:
Here you're calling string(payment.GetMerchantData()) and in GenerateHMACToken you're calling []byte(message).

Can we drop these conversions?

Reply:
There's no real conversion here.  This is zero-cost in Go.